### DISCUSSION TITLE
Keyword _nf Issues

### DISCUSSION THREAD

**Discussion by jbk68** [cortexrd/Knack-Toolkit-Library#216] [General]
Lately I've noticed that the keyword "_nf " is not working at all times as expected.  I have 4 table columns that have successfully hidden from showing up when a users presses Knack's "add filter" button.   But the last couple days, I've noticed that at times, these 4 hidden names will appear in filter drop down list.  If a user changes pages and/or hits refresh, occasionally these names will hide as expected but not always.  Then there are times if the names are hidden in the filtered list and the user refreshes page or changes pages and returns, the 4 names are present again.   ???

Just a fyi, I've tried using both the field name, which is what had been working, and the field id at the view level with same results listed above.  I then tried the keyword at the field level, same result.

Any and all help would be appreciated... JON

**Comment by amandajane-mo** [comment]
Hi Jon,

I've just tested my _nfs and am seeing the same issue. I've tried _nf in both field and view and encountered the same issue. It appears random and no errors appear in console.

We're using Ktl ver 0.22.14, I also tried it in KTL-Dev, with same results.

In our case we're hidding 3 fields from filters, however, it's always just the same 1 field that appears. The issue is intermittent.

Amanda

**Comment by jbk68** [comment]
Hey Amanda, thanks for looking at that for me.  

Don't take this the wrong way but glad to see others are having this issue also and not an isolated issue with me.  :)

Hopefully Norm has a fix for it soon... JON

**Comment by cortexrd** [comment]
Hi guys,

I think I know what's happening.  I apply the kw a bit too early.  As a consequence, some entries are populated afterwards.
I'll see how I can fix this.

Next time, create an issue please.

Norm

**Comment by cortexrd** [comment]
Hi guys,

No matter what I try (_nf in field, view or both), I can never reproduce the bug here.
Is it possible to create a page for me to access and debug in your app?

Norm

**Comment by jbk68** [comment]
Hey Norm, I just added the keyword to the debug page I set up for you a couple weeks ago and the keyword _nf works just fine in that page.  I tried it in the view mode using the actual field id.  Not sure why both me and Amanda are having similar issues in our real live apps.  It weird, when you first open the app and got to add filter the field you want to not show in the list are hidden but when you refresh page or change to another page then back, hit add filter, hidden fields are now shown in drop down list.

Just a fyi, the debug page I setup for you is part of the same app I'm having issue with just setup up a separate page/table named "debug"

Thanks... JON

**Comment by cortexrd** [comment]
@jbk68,

I'm not sure I understand, when you say:

"_I just added the keyword to the debug page I set up for you a couple weeks ago and **the keyword _nf works just fine in that page**._"

Are you saying that I won't be able to reproduce the bug there?

I'd like to test that page, but will I see the bug?

Norm

**Comment by jbk68** [comment]
No you will not see the bug in the "debug page" I set up for you.  If you
remember, the debug page was very simple and only had 4 fields visible on
the table.

I added the keyword _nf to the debug page with the intention of hiding 2 of
the fields (ie _nf=field_1, field_2) and both the fields were successfully
hidden for the filter drop down.  If I removed the keyword, all fields on
the table were shown in the drop down of the filter box.  If I went and
just hid one, that one was hidden.  I tried refreshing the page several
times and could NOT reproduce the issues in my real life app that both me
and Amanda are experiencing.

**Reply by cortexrd** [reply]
Ok then.  I'll try something else.
But it's tough when you're shooting in the dark!

And BTW, I sent you an email.  I can't log in your page anymore.

ND

**Comment by cortexrd** [comment]
@amandajane-mo ,

Any chance I can log in your app to see the bug?

Norm

**Comment by jbk68** [comment]
Norm, trying logging into that debug page now using same credentials.  Just a fyi, I

**Reply by cortexrd** [reply]
...missed a few words?!

**Comment by cortexrd** [comment]
Did someone close this issue?!

**Reply by jbk68** [reply]
I just replied to your email.

**Comment by cortexrd** [comment]
Got it, I'm in now.  I see _nf working as expected, hiding fields 2 and 3.

I've released a **dev** version.  Can you test it to see if it makes a difference?

Norm

**Comment by cortexrd** [comment]
@jbk68,

As expected, I can't reproduce the issue in your app.

Waiting to see what the dev version does for your buggy app...

@amandajane-mo :  if you have the chance to test it also, would be great.

ND

**Comment by jbk68** [comment]
dev mode still shows hidden fields in filter.  How do I get out of dev mode ?

**Comment by cortexrd** [comment]
Gosh... 😞 

To get out of dev mode:  https://github.com/cortexrd/Knack-Toolkit-Library/discussions/186

![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/42851359/82f94e61-f0af-4337-8850-3125bed275bd)

**Comment by jbk68** [comment]
SORRY!!! Wasn't hitting the "cancel" button.

**Reply by cortexrd** [reply]
LOL!!  Don't worry, I often click Ok just by reflex also!

**Comment by cortexrd** [comment]
What about your view's filter settings?  Which mode are you using?
**Used by this view** or **All fields in table**?

![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/42851359/c71553f2-628c-4586-8e17-64ba2324a411)

**Comment by jbk68** [comment]
Using "the fields being used by this view" both in the page that has filtering issues and the "debug" page.

**Reply by cortexrd** [reply]
Good to know.  I tried both and I never see the bug anyways.

**Comment by cortexrd** [comment]
I tried another crude approach just to see if it works, even though it's not a solution.  I added a 200ms delay before cleaning up the filter's list.

So you might see the unwanted fields briefly.

But if they are all removed as expected, and repeatedly, I'll have a good hint on a solution.

Try the **dev** version please,
Norm

**Comment by cortexrd** [comment]
You'll need to do Ctrl+F5 to see the changes.

**Comment by jbk68** [comment]
Same thing, initially the hiding of the fields in filter list works, but if you refresh page, change page and come back, the fields that you want hidden are shown again.

**Comment by amandajane-mo** [comment]
Hi Norm,

Just catching up on this thread. 

Tried in KTL dev, I was able to recreate the issue when _nf on view and on field, grid settings only filter by fields in view. It did take _many_ more attempts to get the errant field to show, but eventually it did, when moving between pages, not after a refresh. 

In my case it's always the same field that shows Description DF, I'm hiding another 2 fields from the filter.

![Screen Shot 2024-02-03 at 15 00 00](https://github.com/cortexrd/Knack-Toolkit-Library/assets/79783211/2e25b389-8b17-4cef-9c02-b546d1edc4b4)

Seems like a slippery one!

Amanda

**Reply by cortexrd** [reply]
Great!  

What is Description DF?  A connected field by any chance?  Or a direct field in the table, just like all others?

ND

**Comment by amandajane-mo** [comment]
It's a direct field in the table, a text formula.

**Comment by jbk68** [comment]
Just a fyi, my fields having issues are currency fields with simple conditions.  I just tried hiding a couple simple text fields and removing the currency field for the equation and had same results.  Works great until page is manually refreshed a couple times or when user moves between pages.  I also noticed that if you open the filter dialog box and close it a couple times, initially the fields are hidden, but will show up again after a couple times of opening/closing filter dialog box.

**Comment by cortexrd** [comment]
Just tried a totally new approach.  May work...

Try dev again please,
Norm

**Comment by amandajane-mo** [comment]
Retried, _nf on field, issue arose after changing pages 5 times. Will now try on view

**Comment by jbk68** [comment]
I retried on both field and view and had same result as Amanda.  Works unit you refresh page multiple times and/or change page and return to page.

**Comment by amandajane-mo** [comment]
Yep, same as Jon field and view. Changing pages and on a refresh.

**Comment by cortexrd** [comment]
I suspect now that the root cause is not even where I'm searching.  I have other ideas, but I need live testing.

Any chance I can connect with one of you on Teams or Zoom to see it for myself?
This would allow adding logs to pinpoint the problem.

I'd do this now.

Or, give me access to your app in any page with the bug.


ND

**Comment by CSWinnall** [comment]
Hi Norm

Just about to go out for dinner can do tomorrow if you're available.

C & A

**Reply by cortexrd** [reply]
Sure will.  This issue is important.
Send me a msg when ready.
Norm

**Comment by cortexrd** [comment]
I've added a few logs.  Let me know what you see compared to mine - **but only when the bug occurs.**

This is the normal expected behavior:  I have **4 fields** in the _nf keyword.

So here's what I see in the console output:

![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/42851359/7a4ad251-3ac2-44bd-8921-1a1ff7143e10)

1. **_nf click processed** should always be seen, wherever you click in the page.
2. If you click on a filter element (Add filters button of in the popup to add new filters), you should see this log: **filter element found**
3. Followed by 4 waiting for: ...
4. Followed by 4 Removing:...

The rest is a sort of _parasitic event_ and not important.

Try the **dev** mode and let me know what is different on your side - when the bug happens.
You can copy/paste your logs here if you want or take screens grabs.

BTW, I've discovered the coolest app for this:  **ShareX**  (https://getsharex.com/).  It allows editing the capture right-away, add text, arrows, boxes, etc.

Norm

**Comment by CSWinnall** [comment]
@cortexrd 

@jbk68 do you have more than 1 grid (of the same object) on the page you see the error? 

We've started to spot a pattern:

1. 2 grids of same object on page, both with same fields to be filtered
4. If we have description DF on both views _nf works as expected
5. When we remove description DF from 1 of the views, it will fail, depending on the order the views are loaded, if the view containing description is loaded first it will fail, the 2nd view (which now doesn't have description in it) appears to determine the filters on the first view.

We're still testing this, but does appear to be a pattern, we tried it with a different field and had the same results.

Amanda

**Reply by jbk68** [reply]
Amanda, on the page have issues, I currently have 2 views but both have a different data source.  One view (upper) is having the issues of showing the (3) fields I wish to hide and the other view (lower) is working fine and is hiding (6) fields successfully.  Never had an issue with lower view.

But after receiving your inquiry I tried the following and saw the (3) fields successfully be hidden regardless of how many times I refreshed page or move between pages.

Here's what I did:  I added to the 2nd (lower) view/grid the (3) fields (connected) that I want to hide in the upper view.  I then at the "field" level, added the KW _nf to the (3) fields having issues.  After doing this, the upper and lower view now BOTH hide the fields in the filtered list.   If I put the KW _nf at the view level, the upper view went back to having issues.  I then add the KW _hc to hide the (3) fields from being viewed on the lower view/grid.

Until you guys get a fix, this is working for me... JON

**Reply by amandajane-mo** [reply]
Mmmm, it's a weird one. Hopefully Norm will be able to track the issue down with our findings.

**Reply by jbk68** [reply]
I have faith he will.  Just a fyi, I had another field from "top view" that I wanted to hid in the filtered list, once I added it the the lower view, and added the kw _nf to field view 's description, it now successfully doesn't appear in either view.

I might add a 3rd view/grid to this page, just to add and then hid these fields, then simply hid the view.   ???

**Comment by CSWinnall** [comment]
@cortexrd 

Screenshots

view_4917 - has description field_6607 on it
view_4961 - doesn't have description

![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/10554243/9250c09b-d523-4ad9-a300-ea7acf42de1c)
In above screenshot description is NOT hidden from filters in view 4917

![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/10554243/7218b761-2860-4299-b200-43a159a7f1b3)
In this case view 4961 loads first and description IS hidden from filters in view 4917

Hope this brings some enlightenment

Amanda

**Comment by cortexrd** [comment]
Ok guy, thanks for your new input.
I can now reproduce the bug at will.

Here's my setup:

A page with only two grids.

Both have Phone and Email columns.

Grid 1 has `_nf=Phone, Email`
Grid 2 has `_nf=Phone`

I can make Grid 1 fail **every time** by doing this sequence:

1. refreshing the page 
2. refreshing the Grid 2 manually with the Click+Click Hold trick

When open the Filters for Grid 1, Phone will be there, though it shouldn't.

Now, let's see how I can fix this...
Norm

**Reply by amandajane-mo** [reply]
Thanks Norm and Good luck

**Reply by jbk68** [reply]
Thank you sir... JON

**Comment by cortexrd** [comment]
Ok, I think I'v got it cornered... again !!!

It was much harder than I thought.

Give it a try please, dev mode.

Norm

**Reply by CSWinnall** [reply]
Hi Norm
Unfortunately, this hasn't worked we are getting this consoled:

![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/10554243/7b8dc029-e0bb-40f2-8e37-42f69ec1ecfb)

And all fields are displayed in the filter

C & A

**Reply by CSWinnall** [reply]
Also we are getting this error:
![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/10554243/af458f12-e83a-4999-859b-cf9249e380c0)

The error happens when we move between pages

**Reply by CSWinnall** [reply]
We are only using the field keyword.  When the error appears and we go back to the page we get the console.log of 
![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/10554243/2374a292-9f3e-4e60-82b6-39eec208515e)

**Reply by cortexrd** [reply]
I can't reproduce this here.

Can you write a **precise configuration and sequence** as I did please?

You can't put too much information!

For example: you should mention that this view has its filter option to **All fields in table**, otherwise it will take me some time before I discover this.

Norm

**Reply by cortexrd** [reply]
Your viewId is undefined.  Can you find out why?

And this line:
`if (Knack.views[viewId].model.view.filter_fields === 'view')`

Can you find out what is the value that is not 'view' ?

ND

**Reply by CSWinnall** [reply]
Here's my setup:

A page with only two grids.

Both tables have same source.  The pages I move between all have 2 tables with the same source.
All Tables have these fields with the _nf keyword:
**Name in Object**: FOR ALL - Task Display Field - **Name in Table**: Audit Name
**Name in Object**: Notes - **Name in Table**: Notes
**Name in Object**: Description DF - **Name in Table**: Description

Page 1 - This is setup for testing nf
Grid 1 has Description, Audit Name
Grid 2 has Audit Name, Notes

All other pages have
Grid 1 - Description, Audit Name, Notes
Grid 2 - Audit Name, Notes

I can recreate the error by:

1. refreshing the page
2. Switching between menu tabs
3. If I don't open the filters the error appears every time I switch pages
4. If I open the filters the error appears once then I get the undefined in console.

I never get the error on a refresh

Hope that explains it better

Craig

**Reply by cortexrd** [reply]
Better, yes.

But what about your Filter option - are they set to **All fields in table** ?

ND

**Reply by CSWinnall** [reply]
![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/10554243/33095264-2fb5-4572-a174-9d8f1b9b8083)

Also all grids filters are set to show only fields being used by this view

**Reply by CSWinnall** [reply]
![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/10554243/377bfe2a-0d40-4826-8a97-3614046dba81)

The screen shots I have posted are all the views that have the afore mentioned setup

**Reply by cortexrd** [reply]
It seems that **view_4916** is problematic.
Can you try this:  **Knack.views['view_4916'].model.view.filter_fields**
That is just another grid like all others?

**Reply by CSWinnall** [reply]
That is the menu view

**Reply by CSWinnall** [reply]
![image](https://github.com/cortexrd/Knack-Toolkit-Library/assets/10554243/12106349-cbf9-4c6f-9a0e-3017cc8a4aeb)

**Reply by cortexrd** [reply]
Ahhhh!  Now I get it.

**Comment by jbk68** [comment]
Hey Norm, I can confirm the keyword _nf is working in dev mode as expected both by selecting the field to hid at field and view levels.

Much appreciated... JON

**Comment by cortexrd** [comment]
@CSWinnall 

Can you reload dev and test please

Norm

**Reply by CSWinnall** [reply]
No Error or console.logs but on switching pages no field are filtered

**Comment by cortexrd** [comment]
No it's worse now...

**Reply by CSWinnall** [reply]
We have given you access to this page:
audits/upcoming-audits/
After the #

**Reply by cortexrd** [reply]
Thanks, I will try that later.
For now, I found another intermittent use case with my setup.  I have to find a solution here before I test your bug.

ND

**Reply by jbk68** [reply]
Just spent the last 15 minutes in dev mode trying different things like refreshing page and switching back and forth thru pages and any and all items that I filter, stay filtered in dev mode.  I even removed the _nf keyword and the fields I want to hide and when I remove the keyword, fields are displayed, put them back in either on the view or field level and filtering works just fine.  

Once I get out of dev mode, filtering issues returns within a few page refreshes.  Dev mode, no issues...

**Comment by cortexrd** [comment]
Ok guys, one more time!

I had to rewrite a big chunk of code with a different approach again.

Please try **dev**.

Norm

**Reply by amandajane-mo** [reply]
Hi Norm,

Sorry, same issue. In our tests it never fails after a refresh but always fails after a move between pages.

Amanda

**Reply by jbk68** [reply]
Still works just fine in dev mode on my end regardless of if I refresh page or move between pages.  I even added a couple extra fields to hide and they were also filtered from being seen in the list.

Not sure what you've done since I initially started this post/issue a few days ago, but ALL my apps are 100% working now, at least in dev mode.

Thank you... JON

**Comment by cortexrd** [comment]
@CSWinnall 

I am on your page now.
I click on all menu items, but can't see your issue.

Shall we have a Teams meeting?

Norm

**Reply by CSWinnall** [reply]
Give us 20 mins?

**Comment by cortexrd** [comment]
Ok, I see the _nf issue.  I was looking for the console error, but it's not there anymore.

**Reply by CSWinnall** [reply]
Yes the error went away.
When changing pages all fields show

**Comment by cortexrd** [comment]
Try **dev** again please?
Norm

**Reply by amandajane-mo** [reply]
We've both tested numerous times and can't get the issue to occur.

Looks like you're on to something

**Comment by cortexrd** [comment]
I will close this and do a release.
If you see something else, reopen.

**Thanks to all for your perseverance and patience!**

Norm

**Reply by jbk68** [reply]
Thanks Norm... JON

**Comment by cortexrd** [comment]
**Released in v0.22.17.**
